Add HTML user interface

Self-Improvement-Lover opened the pull request:
Adds a HTML user interface for the calculator

Self-Improvement-Lover committed:
Add HTML user interface

massimocode commented on index.html:
```diff
@@ -0,0 +1,41 @@
+  </head>
+  <body>
+    <div>
+      <input type="text" id="screen" value="" currentOperation="" disabled />
```
Why do we have a currentOperation attribute, it's not used anywhere

Self-Improvement-Lover replied:
I've removed it

massimocode commented on index.html:
```diff
@@ -0,0 +1,41 @@
+    <br />
+    <button numbersToPress>0</button>
+    <button id="equalsButton">=</button>
+    <button numbersToPress>.</button>
```
Why is "." treated as a number button? The logic for "." is completely different to the logic for any of the other numbers

Self-Improvement-Lover replied:
Good point. I have Fixed that.

massimocode commented on index.js:
```diff
@@ -0,0 +1,138 @@
+const previousAnsButton = document.querySelector("#previousAnswerButton");
+const historyButton = document.querySelector("#getHistory");
+let secondValue = "";
+let operant = "";
```
spelling

Self-Improvement-Lover replied:
Which One?

massimocode replied:
Did you mean operand?

massimocode replied:
The comment is on line 22, it just shows 3 lines of preceeding code for context

massimocode replied:
Should say operand

Self-Improvement-Lover replied:
Alright, did not realise that it was supposed to end with a d and not a t. Thanks, I have updated the spelling.

massimocode commented on index.js:
```diff
@@ -0,0 +1,138 @@
+let currentValue = 0;
+let operantCounter = 0;
+
+const calcualtor = new Calculator();
```
spelling

Self-Improvement-Lover replied:
Changed Spelling

massimocode commented on index.js:
```diff
@@ -0,0 +1,138 @@
+}
+
+clearButton.addEventListener("click", () => {
+  document.getElementById("screen").value = "";
```
You have the screen element in the screen variable, why are you getting it from the DOM again here?

Self-Improvement-Lover replied:
Good point. Changed it.

massimocode commented on index.js:
```diff
@@ -0,0 +1,138 @@
+const operationButtons = document.querySelectorAll("#operationButtons");
+const equalsButton = document.querySelector("#equalsButton");
+const clearButton = document.querySelector("#clearButton");
+const previousAnsButton = document.querySelector("#previousAnswerButton");
```
Why use getElementById in some places and querySelector in others?

Self-Improvement-Lover replied:
Good point, changed it all to querySelector

massimocode commented on index.js:
```diff
@@ -0,0 +1,138 @@
+
+const calcualtor = new Calculator();
+
+function returnValueOfOpperation() {
```
spelling

Self-Improvement-Lover replied:
Changed it.

massimocode commented on index.html:
```diff
@@ -0,0 +1,41 @@
+    <button numbersToPress>1</button>
+    <button numbersToPress>2</button>
+    <button numbersToPress>3</button>
+    <button id="operationButtons">*</button>
```
It's not proper HTML to have multiple elements on the screen with the same ID

Self-Improvement-Lover replied:
Good point. Changed it to being a class.

massimocode commented:
Why does the calculator screen contain the whole operation you are doing? For example:

When I press 9 then 1 then / then 8, the screen says "91/8"

I would expect that when I press the divide button nothing would happen on the screen, then when I press 8, it would just say 8, then when I press equals it would say the result.

Self-Improvement-Lover committed:
Code review changes

Self-Improvement-Lover committed:
updated with massimocode advice

massimocode commented on index.js:
```diff
@@ -0,0 +1,106 @@
+let secondValue = "";
+let operant = "";
+let currentValue = 0;
+let operantPressed = 0;
```
should say operandPressed

massimocode replied:
Also if your intention is for this to be a true/false value, then why do you have it set to 0 and why do you do -- on it? A bit strange really. I would normalise on either true or false, not mix it with numbers

Self-Improvement-Lover replied:
Good point, thanks for bringing that up. Fixed it.

massimocode commented on index.js:
```diff
@@ -0,0 +1,106 @@
+
+const screen = document.querySelector("#screen");
+const numbers = document.querySelectorAll("[numbersToPress]");
+const operationButtons = document.querySelectorAll(".operationButtons");
```
why are the operation buttons selected using classes and the numbersToPress based on attributes?

Also we typically use kebab-case, not camelCase, in HTML attributes and CSS class names

Self-Improvement-Lover replied:
Good point, It would be so much cleaner If they were both classes. 
Alright, changed them to kebab-case.

massimocode commented on index.js:
```diff
@@ -0,0 +1,106 @@
+const screen = document.querySelector("#screen");
+const numbers = document.querySelectorAll("[numbersToPress]");
+const operationButtons = document.querySelectorAll(".operationButtons");
+const equalsButton = document.querySelector("#equalsButton");
```
Why not use document.getElementById?

Self-Improvement-Lover replied:
Is that more recommended to use, since it is more explicit?

massimocode replied:
https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById
If you have an ID then it's quicker and more semantically correct to use this method

Self-Improvement-Lover replied:
OK

massimocode commented on index.js:
```diff
@@ -0,0 +1,106 @@
+const operationButtons = document.querySelectorAll(".operationButtons");
+const equalsButton = document.querySelector("#equalsButton");
+const clearButton = document.querySelector("#clearButton");
+const historyButton = document.querySelector("#getHistory");
```
This isn't needed anymore

Self-Improvement-Lover replied:
True, got rid of it.

massimocode replied:
It's still there

massimocode commented on index.js:
```diff
@@ -0,0 +1,106 @@
+  });
+});
+
+decimalButton.addEventListener("click", () => {
```
Why in this function arr you checking the decimalButton's innerText value? Isn't it always "." anyway? I think this is old code.

massimocode replied:
In fact this function should be simplified greatly, most of the logic is very strange and meaningless at the moment

Self-Improvement-Lover replied:
> d

True, forgot about that. Changed it.

Self-Improvement-Lover replied:
I took out the part you told me about. But to me it seems like its now simplified to the best it can. I am not allowing for multiple "." to be shown on screen value, and if someone pressses the "." first, then the screen value will be "0." which logically makes sense. Or is that what you meant by currently it is meaningless?

massimocode commented on index.js:
```diff
@@ -0,0 +1,106 @@
+
+numbers.forEach((button) => {
+  button.addEventListener("click", () => {
+    if (operantPressed) {
```
This function can be made simpler:
```
numbers.forEach((button) => {
  button.addEventListener("click", () => {
    if (operandPressed) {
      screen.value = "";
      operandPressed = false;
    }
    screen.value += button.innerText;
    secondValue += button.innerText;
  });
});
```

Self-Improvement-Lover replied:
I tried to implement that code and an error occurs. If you press an operand button, and then press a number, then that number won't be shown on screen, since you aren't concatenating it to the screen. Rather, you would have to press the same number twice, for it to appear on the screen once.

massimocode replied:
I think you've copied and pasted the code incorrectly because the code I've put above works fine.

Self-Improvement-Lover replied:
I didn't copy all code becuase i thought it was not needed and only copied a certain part. My mistake. Changed it.

massimocode commented on index.html:
```diff
@@ -0,0 +1,39 @@
+    <button id="equalsButton">=</button>
+    <button id = "decimalButton">.</button>
+    <button class="operationButtons">^</button>
+    <script src="calculator.js"></script>
```
this file tries to set module.exports = Calculator. So above this line please put the following code:
```
    <script>
      const module = {};
    </script>
```

massimocode replied:
That should stop an error from happening

Self-Improvement-Lover replied:
Ok, thanks for that.

massimocode commented:
Another common piece of calculator functionality missing (on most calculators) is the following scenario
1+1+1+1=4
But on yours it does this
1+1+1+1=2

Please fix this, and also have a play with the calculator on your phone or a classical calculator and see what other kinds of behaviour yours still needs.

Self-Improvement-Lover committed:
Able to calculate with multiple operands used

used kebab-case for where appropiate. Improved spelling as well and cleaned code with massimocode advice

massimocode commented on index.html:
```diff
@@ -0,0 +1,42 @@
+<!DOCTYPE html>
```
Please format this document using prettier and push the changes

Self-Improvement-Lover replied:
alright

massimocode commented on index.js:
```diff
@@ -0,0 +1,120 @@
+let secondValue = "";
+let operand = "";
+let currentValue = 0;
+let operandPressed = false; // changed opeant pressed from 0 to false, 1 to true
```
No need for the comment

Self-Improvement-Lover replied:
ok

massimocode commented on index.js:
```diff
@@ -0,0 +1,120 @@
+      break;
+    case "^":
+      calculator.power(currentValue);
+      screen.value = calculator.getTotal();
```
You can have this line once at the bottom of the function. No need to put it in every block

Self-Improvement-Lover replied:
ok

massimocode commented on index.js:
```diff
@@ -0,0 +1,120 @@
+      screen.value = calculator.getTotal();
+      break;
+    default:
+      return currentValue;
```
Return value is never used. This default block isn't necessary.

Self-Improvement-Lover replied:
ok

massimocode commented on index.js:
```diff
@@ -0,0 +1,120 @@
+function returnValueOfOperation() {
+  if (screen.value === "" || screen.value === ".") return;
+
+  if (secondValue.includes(".")) {
```
I think you can always use parseFloat. I'm not aware of there being any scenario where you can't.

Self-Improvement-Lover replied:
ok

massimocode commented on index.js:
```diff
@@ -0,0 +1,120 @@
+let operand = "";
+let currentValue = 0;
+let operandPressed = false; // changed opeant pressed from 0 to false, 1 to true
+let timesOperandUsed = 0
```
This can also be a boolean (true/false) as you don't do anything specific with a value of more than 1.

Self-Improvement-Lover replied:
true, would it be better if i changed it or would it be better if i made it a boolean?

massimocode commented on index.html:
```diff
@@ -0,0 +1,42 @@
+    <button class="operation-buttons">-</button>
+    <br />
+    <button class = "numbers-to-press">0</button>
+    <button id="equals-button">=</button>
```
Whilst CSS classes generally use kebab-case, html IDs generally use camelCase, so equalsButton. This is because they are available as properties on the window object and accessing them like this `window.equalsButton` is easier than `window["equals-button"]`. You can read more about that here https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id

Self-Improvement-Lover replied:
ok

massimocode commented on index.js:
```diff
@@ -0,0 +1,120 @@
+      operand = button.innerText;
+      secondValue = "";
+    
+
```
Please format the files using prettier before committing and pushing

Self-Improvement-Lover replied:
ok

massimocode commented on index.js:
```diff
@@ -0,0 +1,120 @@
+});
+
+decimalButton.addEventListener("click", () => {
+  if (decimalButton.innerText === "." && screen.value.includes(".")) return;
```
You can replace all references of decimalButton.innerText with "."

Self-Improvement-Lover replied:
I see your point

massimocode commented on index.js:
```diff
@@ -0,0 +1,120 @@
+
+decimalButton.addEventListener("click", () => {
+  if (decimalButton.innerText === "." && screen.value.includes(".")) return;
+  if (screen.value === "") {
```
Why would screen.value ever be ""? Have you ever seen a calculator that you turn on and it doesn't just say "0"? I think your calculator should be the same.

Self-Improvement-Lover replied:
Good point, I have changed that.

massimocode commented:
If I press 5 + 5 + + , on the last + button I get ERROR

Self-Improvement-Lover committed:
Updated with automatic 0 screen and can doing multiple straight additions with getting wrong answer

Self-Improvement-Lover committed:
added API and improved minor bugs

Self-Improvement-Lover committed:
fully-client-side-calculator

Self-Improvement-Lover committed:
Improved File Naming & API

Self-Improvement-Lover committed:
Updates File Naming & Sorting and better API

Self-Improvement-Lover committed:
without web-based calculator

Self-Improvement-Lover committed:
.gitignore working

massimocode commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+web-based-calculator 
+node_modules
```
doesn't need to be here, as there's no package json or node modules at the root

Self-Improvement-Lover replied:
Can you explain further? node_modules is still in the calculate server folder, shouldn't i still mention this in .gitignore, so that it doesn't get pushed up to github? or is it when it is not at the root, it is automatically not pushed up?

massimocode replied:
They must be on your disk only. They aren't here. You can delete them.

massimocode commented on calculator-server/app.js:
```diff
@@ -0,0 +1,11 @@
+app.use(cors())
+app.use(bodyParser.json())
+app.use('/api/v1/calculator', router)
+app.listen(port, console.log('Server is up'))
```
You're technically calling `Server is up` before calling `app.listen`, let alone before `app.listen` actually finishes.

Also what if there's an error listening? You're still logging "server is up".

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+const Calculator = require("../../core/calculator");
```
please format all code before submitting for review

Self-Improvement-Lover replied:
hopefully

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+  console.log(req.body);
+  const { operand1, operand2, operation } = req.body;
+
+  const calculator = new Calculator();
```
Work should be done after the request has been validated

Self-Improvement-Lover replied:
Alright, done

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+  calculator.add(operand1);
+
+if (operation === undefined){
+  res.status(400).json({ error: "An operation is needed" });
```
you should have a return statement after this line to exit the function and not continue trying to process the request

Self-Improvement-Lover replied:
True, done

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+}
+
+if ((typeof operand1 !== 'number' )|| (typeof operand2 !== 'number' ) ){
+  res.status(400).json({ error: "Please enter numbers as the operands" });
```
Same here, return

Self-Improvement-Lover replied:
Alright

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+}
+
+if (!(operand1 && operand2)){
+  res.status(400).json({ error: "Please enter 2 operands" });
```
Same here, return

Self-Improvement-Lover replied:
Alright, done

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+  res.status(400).json({ error: "Please enter numbers as the operands" });
+}
+
+if (!(operand1 && operand2)){
```
If an operand was not provided, then of course it's not going to be a number, as per the above validation, so this validation needs to be done before that one

Self-Improvement-Lover replied:
True, fixed that

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+  res.status(400).json({ error: "Please enter numbers as the operands" });
+}
+
+if (!(operand1 && operand2)){
```
This will not allow a user to provide the operand 0

Self-Improvement-Lover replied:
True, I fixed that

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+    calculator.power(operand2)
+    break;
+  default: 
+  res.status(400).json({ error: "Not an accepted operation/ not an operation" });
```
You need to return when there's an error

Self-Improvement-Lover replied:
True, fixed that

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,67 @@
+if (total === "ERROR") {
+  res.status(400).json({ error: "Cannot divide by Zero" });
+} else {
+  res.status(200).json({ result: total });
```
Get rid of the else and put a return in the if above

Self-Improvement-Lover replied:
Done

massimocode commented on calculator.js:
This file should not have been deleted. I think you moved it to core. So check that in.

Self-Improvement-Lover replied:
It is in the core folder now.

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,124 @@
+let secondValue = "";
+let operand = "";
+let currentValue = 0;
+let operantPressed = false;
```
operand*************

Self-Improvement-Lover replied:
Changed it

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,124 @@
+const calculator = new Calculator();
+
+function returnValueOfOperation() {
+  if (screen.value === "" || screen.value === ".") return;
```
braces and format

Self-Improvement-Lover replied:
Done

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,124 @@
+
+    // if when second number entered is at 0., then number button is pressed, we dont want screen value to be reset,
+    //  we want the number button pressed to be concatnated to the 0.
+    if (operantPressed && screen.value === "0.") {
```
I think if the screen value is 0., then operandPressed will never be true, so this section of code can never run and is redundant

Self-Improvement-Lover replied:
If you press . then screen value updates to 0. and not just . since no calculator does that.

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,124 @@
+const calculator = new Calculator();
+
+function returnValueOfOperation() {
+  if (screen.value === "" || screen.value === ".") return;
```
I think screen.value can never be ".", so this second section of the condition is not required.

Self-Improvement-Lover replied:
True

Self-Improvement-Lover committed:
Fixed erros in calculator API and Formatted Code

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,65 @@
+    return;
+  }
+
+  if (operand1 !== 0 && operand2 !== 0 && !(operand1 || operand2)) {
```
`if (operand1 === undefined || operand2 === undefined) {`

massimocode commented on calculator-server/controllers/calculator-stateless.js:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  const total = calculator.getTotal();
+  // have to wait for operation to occur, before we can get total
```
This comment can be removed

Self-Improvement-Lover replied:
ok

massimocode commented on calculator-server/package.json:
You have a .gitignore missing for node_modules inside the calculator-server folder

massimocode commented on calculator-server/package.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "calculator-web-page",
```
calculator-server

Self-Improvement-Lover committed:
removed error where you can not cocatnate 0. to any other number

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,141 @@
+const clearButton = document.getElementById("clearButton");
+const decimalButton = document.getElementById("decimalButton");
+let secondValue = "";
+let operand = "";
```
Rename to `operation`

Self-Improvement-Lover replied:
OK

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,141 @@
+
+numbers.forEach((button) => {
+  button.addEventListener("click", () => {
+    // if first number entered is 0. something,then an operandButton is pressed,then number,
```
Remove the comment and commented out code if you don't need it

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,141 @@
+
+operandButtons.forEach((button) => {
+  button.addEventListener("click", () => {
+    // if an operand is pressed twice or more times in a row
```
This comment should be moved down a line

Self-Improvement-Lover replied:
OK

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,141 @@
+    return;
+  }
+
+  screen.value = screen.value + decimalButton.innerText;
```
use "." instead of decimalButton.innerText, for the line below too

massimocode commented on fully-client-side-calculator/index.js:
```diff
@@ -0,0 +1,141 @@
+}
+
+clearButton.addEventListener("click", () => {
+  screen.value = "0";
```
It would be good to have a clear function defined that did this. The clear button should call the clear function and then also call calculator.clear() separately.

Then you would call the clear function above instead of initialising these variables manually. You would just do `let blah, blah2, blah3;`

Self-Improvement-Lover replied:
I have created the clear function as you have instructed, however I don't understand what you mean by Initialising these variables manually by doing let blah, blah2, blah3? If I did that then would the variables not be accessible to the rest of the functions that are being used due to scope rules?

Self-Improvement-Lover committed:
better variable name and adding new function for efficency
